### PR TITLE
net: openthread: Implement energy scan 

### PR DESCRIPTION
**Pull request by MarekPorwisz**
- Extended radio API to include "energy scan" capability
- Extended nrf5 radio shim to support energy scan feature
- Extended open thread shim to make use of energy scan feature.

tested on nRF52840

### CODE REVIEW

File: include/net/ieee802154_radio.h [hunk 1]
```diff
@@ -26,14 +26,17 @@ extern "C" {
  * @{
  */
 
+typedef void (*energy_scan_done_cb)(struct device *dev, s16_t max_ed);
```
Comment (tbursztyka) [inline include/net/ieee802154_radio.h]: s/energy_scan_done_cb/energy_scan_done_cb_t

File: drivers/ieee802154/ieee802154_nrf5.c [hunk 2]
```diff
@@ -45,6 +45,7 @@ struct nrf5_802154_config {
 };
 
 static struct nrf5_802154_data nrf5_data;
+static energy_scan_done_cb energy_scan_done;
```
Comment (tbursztyka) [inline drivers/ieee802154/ieee802154_nrf5.c]: move it _into_ struct nrf5_802154_data

File: drivers/ieee802154/ieee802154_nrf5.c [hunk 3]
```diff
@@ -183,6 +185,26 @@ static int nrf5_set_channel(struct device *dev, u16_t channel)
...
 }
 
+static int nrf5_energy_scan_start(struct device *dev,
+				  u16_t duration,
+				  energy_scan_done_cb done_cb)
+{
+	int err = 0;
+
+	ARG_UNUSED(dev);
+	if (energy_scan_done == NULL) {
```
Comment (tbursztyka) [inline drivers/ieee802154/ieee802154_nrf5.c]: have an empty line here

File: drivers/ieee802154/ieee802154_nrf5.c [hunk 4]
```diff
@@ -183,6 +185,26 @@ static int nrf5_set_channel(struct device *dev, u16_t channel)
...
+		energy_scan_done = done_cb;
+		if (nrf_802154_energy_detection(duration * 1000) == false) {
```
Comment (tbursztyka) [inline drivers/ieee802154/ieee802154_nrf5.c]: same

(it's a linux rule that zephyr follows btw)

**Review by tbursztyka — CHANGES_REQUESTED** [review]

File: include/net/ieee802154_radio.h [hunk 5]
```diff
@@ -26,14 +26,17 @@ extern "C" {
  * @{
```
Comment (rlubos) [inline include/net/ieee802154_radio.h]: nit: Please update the commit title prefix: `net: ieee802154:`

File: include/net/ieee802154_radio.h (hunk 1 above)
Comment (MarekPorwisz) [reply include/net/ieee802154_radio.h]: renamed

File: drivers/ieee802154/ieee802154_nrf5.c (hunk 2 above)
Comment (MarekPorwisz) [reply drivers/ieee802154/ieee802154_nrf5.c]: moved

File: drivers/ieee802154/ieee802154_nrf5.c (hunk 3 above)
Comment (MarekPorwisz) [reply drivers/ieee802154/ieee802154_nrf5.c]: added empty line before if

File: drivers/ieee802154/ieee802154_nrf5.c (hunk 4 above)
Comment (MarekPorwisz) [reply drivers/ieee802154/ieee802154_nrf5.c]: added empty line before if

File: include/net/ieee802154_radio.h (hunk 5 above)
Comment (MarekPorwisz) [reply include/net/ieee802154_radio.h]: updated comment

**Review by rlubos — APPROVED** [review]
LGTM 👍, thanks!

**Review by tbursztyka — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

File: subsys/net/lib/openthread/platform/radio.c [hunk 6]
```diff
@@ -337,11 +414,26 @@ void otPlatRadioSetPromiscuous(otInstance *aInstance, bool aEnable)
...
-	ARG_UNUSED(aScanDuration);
+	energy_detection_time    = aScanDuration;
+	energy_detection_channel = aScanChannel;
+
+	clear_pending_events();
+
+	radio_api->set_channel(radio_dev, aScanChannel);
 
-	return OT_ERROR_NOT_IMPLEMENTED;
+	if (radio_api->ed_scan(radio_dev, energy_detection_time,
```
[the remark this answers is not in the document]
Comment (MarekPorwisz) [reply subsys/net/lib/openthread/platform/radio.c]: Function description states: 
"This function is used when radio provides OT_RADIO_CAPS_ENERGY_SCAN capability."

and OT code confirms that:
"    if (RadioSupportsEnergyScan()) <- tests the capability
    {
        Get<Radio>().EnergyScan(aScanChannel, aScanDuration); <- calls energy scan
        SetState(kStateEnergyScan);
    }"

**Comment by jhedberg** [comment]
@jukkar looks like merging this broke CI?

https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/65192/1/console

**Comment by jhedberg** [comment]
@jukkar created a hotfix here: #23377

**Pull request merged**